### PR TITLE
Deserialization: zero-copy merge subframes when possible

### LINKED ISSUE #5107: Unnecessary deep copy causes memory flare on network comms
distributed git tip, Linux x64

```python
import distributed
import numpy

c = distributed.Client(n_workers=4, threads_per_worker=1, memory_limit="2.8 GiB")
f = c.submit(numpy.random.random, 2 ** 27)  # 1 GiB
c.replicate(f, 4)
# Alternative to replicate, identical effect
# futures = [c.submit(lambda x: None, f, pure=False, workers=[w]) for w in c.has_what()]
```
## Expected behaviour
Thanks to pickle 5 buffers, the peak RAM usage on each worker is 1 GiB

## Actual behaviour
I can see on the dashboard the RAM of all workers that receive the computed future over the network briefly flare up to 2 GiB and then settle down at 1 GiB.
On stderr I read:
```
distributed.worker - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker.html#memtrim for more information. -- Unmanaged memory: 2.07 GiB -- Worker memory limit: 2.80 GiB
distributed.worker - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker.html#memtrim for more information. -- Unmanaged memory: 2.07 GiB -- Worker memory limit: 2.80 GiB
distributed.worker - WARNING - Unmanaged memory use is high. This may indicate a memory leak or the memory may not be released to the OS; see https://distributed.dask.org/en/latest/worker.html#memtrim for more information. -- Unmanaged memory: 2.07 GiB -- Worker memory limit: 2.80 GiB
```
If I reduce the memory_limit to 2 GiB, the workers get killed off.

The sender worker is unaffected by the flaring.

I tested on Python 3.8 and 3.9 and on protocols tcp://, ws:// and ucx:// and all are equally affected.

### PR DESCRIPTION
**Pull request by gjoseph92**
Better implementation of #5112.

The point of this change (copied from original PR; note I tested and the screencap behavior works with this PR too):

> When merging sub-frames before deserializing, if all the sub-frames are memoryviews backed by the same underlying buffer, "merge" them without copying by making a new memoryview pointing to the same buffer, instead of always copying.
> 
> Running @crusaderky's reproducer from https://github.com/dask/distributed/issues/5107 on `main`:
> before
> With this PR:
> after

I'd always wanted to merge the memoryviews by comparing their pointers, but couldn't find an API to access that information, hence the gnarly previous PR passing around byte offsets.

Thanks to @jakirkham on https://github.com/dask/distributed/pull/5112#issuecomment-890280886, I wrote the pointer arithmetic using NumPy to access the buffer interface. Then I found @mattgwwalker's great post explaining how to do exactly what we needed just using `ctypes`: https://mattgwwalker.wordpress.com/2020/10/15/address-of-a-buffer-in-python/.

- [x] Closes #5107
- [x] Tests added / passed
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

### CODE REVIEW

File: distributed/protocol/serialize.py
Comment (crusaderky) [inline distributed/protocol/serialize.py:490]: The whole block 477:490 could be replaced by
```python
subframes = frames[offset : offset + n]
try:
    merged = merge_memoryviews(subframes)
except (ValueError, TypeError):
    merged = bytearray().join(subframes)
```
after small changes to merge_memoryviews shown below.

**Comment by jakirkham** [comment]
Neat that's a clever way to access pointers in Python 😄 Will take a closer look (just got back and am catching up 😅)

cc @madsbk (as you may have thoughts on zero-copy frame combination)

cc @quasiben (for general awareness)

File: distributed/protocol/utils.py
Comment (gjoseph92) [reply distributed/protocol/utils.py:3]: I didn't know `collections.abc` was preferred now with postponed annotations instead of `typing`

File: distributed/protocol/utils.py
Comment (gjoseph92) [reply distributed/protocol/utils.py:117]: I was trying to make the mico-optimization of not using an exception for control flow in this case (which is common for non-TCP comms), but I don't think that's worth it, and I agree it's nicer the way you've proposed.

File: distributed/protocol/utils.py
Comment (gjoseph92) [reply distributed/protocol/utils.py:185]: I figured the error message was self-explanatory and didn't need the ImportError chained onto it, why would you prefer to keep it?

Comment (crusaderky) [reply distributed/protocol/utils.py:192]: ``Exception.__cause__`` is only set if you explicitly ask for it, e.g.

```python
try:
    ....
except Exception1 as e:
    raise Exception2(...) from e
```
Writing ``raise ... from None`` is redundant as it is the default.

Comment (gjoseph92) [reply distributed/protocol/utils.py:192]: `__cause__` is None in either case, but when you explicitly set it to None, that also sets `__suppress_context__` to True, so the traceback is displayed as only the last exception. It's just a minor style thing; I think it's nice to shorten the displayed traceback when possible, so it's easier to understand if a user sees it. It's all irrelevant in this case anyway, since this error will always be swallowed in `merge_and_deserialize`.

```python
In [4]: try:
   ...:     import asdf
   ...: except ImportError:
   ...:     raise ValueError("couldn't import")
   ...: 
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
<ipython-input-4-d39ad270c21f> in <module>
      1 try:
----> 2     import asdf
      3 except ImportError:

ModuleNotFoundError: No module named 'asdf'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-4-d39ad270c21f> in <module>
      2     import asdf
      3 except ImportError:
----> 4     raise ValueError("couldn't import")
      5 

ValueError: couldn't import

In [5]: try:
   ...:     import asdf
   ...: except ImportError:
   ...:     raise ValueError("couldn't import") from None
   ...: 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-28105f21527c> in <module>
      2     import asdf
      3 except ImportError:
----> 4     raise ValueError("couldn't import") from None
      5 

ValueError: couldn't import
```
```python
In [2]: try:
   ...:     try:
   ...:         import asdf
   ...:     except ImportError:
   ...:         raise ValueError("couldn't import")
   ...: except ValueError as e:
   ...:     print(f"{e.__cause__=}, {e.__context__=}, {e.__suppress_context__=}")
   ...: 
   ...: 
e.__cause__=None, e.__context__=ModuleNotFoundError("No module named 'asdf'"), e.__suppress_context__=False

In [3]: 

In [3]: try:
   ...:     try:
   ...:         import asdf
   ...:     except ImportError:
   ...:         raise ValueError("couldn't import") from None
   ...: except ValueError as e:
   ...:     print(f"{e.__cause__=}, {e.__context__=}, {e.__suppress_context__=}")
   ...: 
e.__cause__=None, e.__context__=ModuleNotFoundError("No module named 'asdf'"), e.__suppress_context__=True
```

**Review by crusaderky — APPROVED** [review]

**Comment by gjoseph92** [comment]
@jakirkham do you have any more thoughts on this? @jcrist mentioned he thought frame splitting itself might be unnecessary except for websockets, and should be handled at the transport level instead.

I think this PR has become somewhat low priority because there's discussion around significantly reworking the protocol, but I do have a feeling it would make a difference for users right now.

**Comment by jakirkham** [comment]
I think this is still useful. Have some WIP review comments that I've fallen behind. Will try to catch back up later this week or next week.

Yeah at least on the `read` side we are allocating one big buffer, which we read into. Agree there's some legacy splitting in serialization that would be good to revisit. IIRC compression is one of the use cases for splitting during serialization (not sure if that is affecting us in deserialization as well).

To Jim's point, I think only frame splitting we do on for communication is baked into the transport layers already. One case is websockets. Another is working around a bug in OpenSSL 1.0.2. In either case neither of these should be in serialization. Jim if you see other issues please let us know.

**Comment by fjetter** [comment]
I think this is a high impact issue. Are there any major concerns left before we can merge? I get that some parts of the protocol are currently in review and might change but until then this should deliver a significant improvement. The changes are well enough isolated that we can easily revert if necessary.

friendly ping @jakirkham @madsbk @jcrist 

If there are no objections within the next 24h I will go ahead and merge.

**Comment by jcrist** [comment]
I've given this a read through and it seems fine to me to merge as is. I still think that splitting large buffers should be done at the transport level instead of here, but for now this seems fine. Test failure is unrelated, should should be good-to-go.

**Comment by jakirkham** [comment]
Have some draft review comments above that I haven't had a chance to finish. Can try to finish them, but unfortunately it won't happen for a few more days. That said, feel free to merge anyways if you need this soon and can try to follow up later

**Comment by gjoseph92** [comment]
@fjetter shall we merge this then? I'd love to see it in the next release.

**Review by jakirkham — COMMENTED** [review]
Apologies for the delay. Had a few minor comments below

File: distributed/protocol/utils.py
Comment (gjoseph92) [reply distributed/protocol/utils.py:174]:
```python
In [18]: %timeit ctypes.c_byte * 1
109 ns ± 1.32 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```
seemed cheap enough I wasn't worried about it, but might as well

**Review by jakirkham — APPROVED** [review]

**Comment by gjoseph92** [comment]
This is still open. I've merged main and the test failures feel flaky to me, but I'm not sure.

**Comment by jakirkham** [comment]
FWIW merged @pentschev's PR ( https://github.com/dask/distributed/pull/5484 ), which included some test fixes for UCX that should clear out the gpuCI failures seen here. May need to merge again to get those

Did see a couple unrelated test failures over there as well. Maybe @jrbourbeau will be able to let us know if these are flaky tests or if they need further investigation

**Review by madsbk — APPROVED** [review]
Nice work @gjoseph92, clever approach that is useful even if we move the splitting/merging to the transport layer!

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
gpuCI now passed, thanks for checking this @jakirkham !

**Comment by gjoseph92** [comment]
This is green besides `test_reschedule_concurrent_requests_deadlock` failing on all Windows machines. Also `test_aliases_2` failed on macOS, which seems odd to me to be flaky, definitely doesn't seem related.

**Comment by jrbourbeau** [comment]
Apologies for the delay here. A few folks have approved this PR, so I'll plan to merge this in later today if no further comments

**Comment by gjoseph92** [comment]
Thanks @jrbourbeau and everyone for reviewing! Very excited to have this in.

**Pull request merged**
